pkg/proc/native/proc.go

```diff
@@ -333,20 +335,48 @@ func (dbp *nativeProcess) handlePtraceFuncs() {
 	runtime.LockOSThread()
 
 	for fn := range dbp.ptraceChan {
+		dbp.insidePtraceThread = true
 		fn()
+		dbp.insidePtraceThread = false
 		dbp.ptraceDoneChan <- nil
 	}
+	close(dbp.ptraceDoneChan)
 }
 
 func (dbp *nativeProcess) execPtraceFunc(fn func()) {
+	if dbp.insidePtraceThread {
```

Comment:
Maybe instead of checking this state we store the thread id of the "magic" thread and then compare here. Something like `if dbp.magicThread == syscall.Gettid()`, I wonder if we lose out on the performance improvements by injecting a syscall in this path.

Otherwise I fear we're opening ourselves to bugs where this global state causes other goroutines to attempt to run ptrace operations due to programmer error.

Reply:
I just tried it and it's actually fast. However I don't think this will help.

The problem only arises when an Interloper-goroutine runs in parallel with the Continue-goroutine and tries to do a ptrace call. It stands to reason that the Continue-goroutine will always have to synchronize with the Interloper-goroutine, if it didn't the Interloper-goroutine could do ptrace calls *after* the Continue-goroutine has resumed the target process, which would fail (here I'm assuming that the Interloper-goroutine will limit itself to reading/writing memory, i.e. that we will never have two goroutines trying to simultaneously manage the run state of the target process, I think this is reasonable).

If we stored and checked the thread id of the magic thread (which would be the thread executing the Continue-goroutine) then the Interloper-goroutine would try to send on ptraceChan, but ptraceChan wouldn't be read until the Continue-goroutine terminates, and since the Continue-goroutine needs to wait for the Interloper-goroutine before resuming the thread we would always get a deadlock.

We could check if the thread ID is the same as the ExecOnMagicThread thread ID and return an error if it isn't, I think a runtime error is marginally preferable to a deadlock, but the OS is already doing this for us (it's ESRCH on linux).

And there's also the fact that Gettid doesn't exist anywhere else but linux. In fact from a quick look to the syscall package I can't figure out how to do this on freebsd and windows.